File: global/expo-localtunnel/ios/ExpoLocaltunnelModule.swift
import ExpoModulesCore
import Network

public class ExpoLocaltunnelModule: Module {
  private var connections: [String: TunnelConnection] = [:]
  private let connectionsLock = NSLock()
  private let queue = DispatchQueue(label: "expo.localtunnel", qos: .userInitiated)

  /// Synchronous wrapper around `connectionsLock`. The helper itself is non-`async`,
  /// so the `noasync`-annotated `NSLock.lock()`/`unlock()` calls inside are reachable
  /// from `AsyncFunction` closures without tripping Swift 6 concurrency diagnostics.
  private func withConnectionsLock<T>(_ body: () -> T) -> T {
    connectionsLock.lock()
    defer { connectionsLock.unlock() }
    return body()
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoLocaltunnel")

    Events(
      "onConnectionOpen",
      "onConnectionClose",
      "onConnectionError",
      "onConnectionDead",
      "onRequest"
    )

    OnDestroy {
      let snapshot = self.withConnectionsLock { () -> [String: TunnelConnection] in
        let snap = self.connections
        self.connections.removeAll()
        return snap
      }
      for (_, conn) in snapshot {
        conn.close()
      }
    }

    AsyncFunction("createTunnelConnection") { (connectionId: String, config: TunnelConnectionConfig) in
      // Clean up existing connection with this ID if any
      let existing = self.withConnectionsLock { self.connections[connectionId] }
      existing?.close()

      let conn = TunnelConnection(
        id: connectionId,
        config: config,
        queue: self.queue
      ) { (name: String, body: [String: Any]) in
        self.sendEvent(name, body)
      }
      self.withConnectionsLock { self.connections[connectionId] = conn }
      try await conn.connectRemote()
    }

    AsyncFunction("closeTunnelConnection") { (connectionId: String) in
      let conn = self.withConnectionsLock { self.connections.removeValue(forKey: connectionId) }
      conn?.close()
    }

    AsyncFunction("closeAllTunnelConnections") {
      let snapshot = self.withConnectionsLock { () -> [String: TunnelConnection] in
        let snap = self.connections
        self.connections.removeAll()
        return snap
      }
      for (_, conn) in snapshot {
        conn.close()
      }
    }
  }
}

struct TunnelConnectionConfig: Record {
  @Field var remoteHost: String = ""
  @Field var remotePort: Int = 0
  @Field var localHost: String = "localhost"
  @Field var localPort: Int = 0
  @Field var localHostHeader: String?
}

File: global/expo-localtunnel/android/src/main/java/expo/modules/localtunnel/ExpoLocaltunnelModule.kt
package expo.modules.localtunnel

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.util.concurrent.ConcurrentHashMap

class TunnelConnectionConfig : Record {
  @Field val remoteHost: String = ""
  @Field val remotePort: Int = 0
  @Field val localHost: String = "localhost"
  @Field val localPort: Int = 0
  @Field val localHostHeader: String? = null
}

class ExpoLocaltunnelModule : Module() {
  // ConcurrentHashMap: AsyncFunction handlers (JS dispatch thread) and event
  // callbacks from TunnelConnection (connection-internal threads) both touch
  // this map. Per-key operations are atomic; we use compute/remove for any
  // compound (read-then-write) operations to keep them atomic too.
  private val connections = ConcurrentHashMap<String, TunnelConnection>()

  override fun definition() = ModuleDefinition {
    Name("ExpoLocaltunnel")

    Events(
      "onConnectionOpen",
      "onConnectionClose",
      "onConnectionError",
      "onConnectionDead",
      "onRequest"
    )

    OnDestroy {
      // Snapshot then clear: drain entries atomically so we don't miss or
      // double-close any concurrently-added connection.
      val snapshot = ArrayList(connections.values)
      connections.clear()
      snapshot.forEach { it.close() }
    }

    AsyncFunction("createTunnelConnection") { connectionId: String, config: TunnelConnectionConfig ->
      val newConn = TunnelConnection(
        id = connectionId,
        config = config
      ) { name, body -> sendEvent(name, body) }
      // Atomically replace any existing entry for this id. The previous value
      // (if any) is closed after the swap so handlers always see the new one.
      val previous = connections.put(connectionId, newConn)
      previous?.close()
      newConn.connectRemote()
    }

    AsyncFunction("closeTunnelConnection") { connectionId: String ->
      // remove() returns the previous mapping atomically.
      connections.remove(connectionId)?.close()
    }

    AsyncFunction("closeAllTunnelConnections") {
      val snapshot = ArrayList(connections.values)
      connections.clear()
      snapshot.forEach { it.close() }
    }
  }
}
